File: 0079-word-search/0079-word-search.cpp
class Solution {
public:
    int directions[4][2] = {{1,0},{-1,0},{0,1},{0,-1}};
    // bool ans = false;
    bool exist(vector<vector<char>>& board, string word)
    {
        int n = board.size(), m = board[0].size();
        vector<vector<bool>> visited(n, vector<bool>(m, false));
        bool ans = false;
        for(int i = 0; i < n; i++)
        {
            for(int j = 0; j < m; j++)
            {
                if(board[i][j] == word[0])
                {
                    visited[i][j] = true;
                    ans |= dfs(i, j, n, m, 1, word.length(), word.substr(0,1), word, board, visited);
                    visited[i][j] = false;
                }
            }
        }
        
        return ans;
    }
    
    bool dfs(int x, int y, int n, int m, int start, int size, string curt, string target, vector<vector<char>>& board, vector<vector<bool>>& visited)
    {
        if(start == size)
        {
            if(curt == target)
            {
                return true;
            }
            return false;
        }
        
        if(target.find(curt) != 0) return false;
        
        
        for(auto direction : directions)
        {
            int new_x = x + direction[0], new_y = y + direction[1];
            if(new_x < 0 || new_x >= n || new_y < 0 || new_y >= m) continue;
            if(visited[new_x][new_y]) continue;
            
            if(board[new_x][new_y] == target[start])
            {
                visited[new_x][new_y] = true;
                curt += board[new_x][new_y];
                if(dfs(new_x, new_y, n, m, start + 1, size, curt, target, board, visited)) return true;
                curt = curt.substr(0, curt.length() - 1);
                visited[new_x][new_y] = false;
            }
        }
        
        return false;
        
    }
};

File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        directions = [[0,1],[0,-1],[1,0],[-1,0]]
        dim = [len(board), len(board[0])]
        
        def recur(x, y, curt, pos):
            if len(curt) >= len(word):
                if curt == word:
                    return True
                return False
            
            ret = False
            
            for direction in directions:
                newx = x + direction[0]
                newy = y + direction[1]
                
                if newx < 0 or newx >= dim[0] or newy < 0 or newy >= dim[1]:
                    continue
                
                if pos < len(word) and board[newx][newy] == word[pos]:
                    temp = board[newx][newy]
                    board[newx][newy] = ""
                    ret |= recur(newx, newy, curt + word[pos], pos + 1)
                    board[newx][newy] = temp
            
            return ret
        
        
        for i in range(dim[0]):
            for j in range(dim[1]):
                if board[i][j] == word[0]:
                    temp = board[i][j]
                    board[i][j] = ""
                    ret = False
                    ret |= recur(i, j, word[0], 1)
                    if ret:
                        return True
                    board[i][j] = temp
        return False